### PR TITLE
[semantic-python] Compile function calls.

### PR DESCRIPTION
**Pull request by patrickt**
- [x] Depends on #320.
- [x] Depends on #322.

Though this does not work with keyword arguments, list/dictionary
splats, or generator expressions, it suffices for the simple case.

### CODE REVIEW

File: semantic-python/src/Language/Python/Core.hs [hunk 1]
```diff
@@ -203,7 +203,18 @@ instance Compile Py.Block where
...
 instance Compile Py.BreakStatement
-instance Compile Py.Call
+
+instance Compile Py.Call where
+  compileCC it@Py.Call
+    { function
+    , arguments = L1 Py.ArgumentList { extraChildren = args }
+    } cc = do
+    func <- compileCC function cc
+    -- Python function arguments are defined to evaluate left to right.
```
Comment (robrix) [inline semantic-python/src/Language/Python/Core.hs]: ❤️

File: semantic-python/src/Language/Python/Core.hs [hunk 2]
```diff
@@ -203,7 +203,18 @@ instance Compile Py.Block where
...
+    args <- traverse (flip compileCC cc) args
```
Comment (robrix) [inline semantic-python/src/Language/Python/Core.hs]: This (and line 212) is wrong:

1. We definitely shouldn’t transfer control to the arguments (passing `cc` to them).
2. We _definitely_ definitely shouldn’t transfer control more than once (passing `cc` to multiple things).

We want to use the continuation once and only once, after `func $$* args`.

Instead, we want to compile all of these subterms as expressions—i.e. give them a “pure” continuation that simply returns their values. I’m not sure if we finished that approach?

File: semantic-python/src/Language/Python/Core.hs [hunk 3]
```diff
@@ -278,9 +290,17 @@ instance Compile Py.IfStatement where
 instance Compile Py.ImportFromStatement
 instance Compile Py.ImportStatement
 instance Compile Py.Integer
+
+-- Buggy: ignores the semantics of the provided name. This requires
+-- coordination between Call nodes and KeywordArgument nodes, which is
+-- a sign that we need something smarter here, but it'll do for now.
```
Comment (robrix) [inline semantic-python/src/Language/Python/Core.hs]: I think it suggests we shouldn’t have a `Compile` instance for `Py.KeywordArgument`, and should instead pattern match on it in the `Py.Call` instance.

File: semantic-python/src/Language/Python/Core.hs (hunk 2 above)
Comment (robrix) [reply semantic-python/src/Language/Python/Core.hs]: (To illustrate the wrongness, try a test case where you do something after a function call with multiple arguments.)

Comment (patrickt) [reply semantic-python/src/Language/Python/Core.hs]: This should be fixed now with the power gained from #320.

File: semantic-python/src/Language/Python/Core.hs [hunk 4]
```diff
@@ -196,7 +196,22 @@ instance Compile Py.Block where
...
+
+instance Compile Py.Call where
+  compile it@Py.Call
+    { function
+    , arguments = L1 Py.ArgumentList { extraChildren = args }
+    } cc next = do
+    func <- compile function pure next
+    let compileArg = \case
+          Prj expr -> compile (expr :: Py.Expression Span) pure next
+          other   -> fail ("Can't compile non-expression function argument: " <> show other)
```
Comment (patrickt) [inline semantic-python/src/Language/Python/Core.hs]: Both the function and its arguments are evaluated in a pure context. The continuation is invoked once the latter have been applied to the former.

Comment (robrix) [reply semantic-python/src/Language/Python/Core.hs]: This is superb.

**Comment by robrix** [comment]
Accidentally merged #331 into this instead of this into `master` and then #331 into `master`. Let’s merge this once green 😅

**Pull request merged**